src/cargo/util/config.rs

```diff
@@ -675,14 +700,19 @@ fn walk_tree<F>(pwd: &Path, mut walk: F) -> CargoResult<()>
     where F: FnMut(File, &Path) -> CargoResult<()>
 {
     let mut current = pwd;
+    let mut stash: HashSet<ConfigFile> = HashSet::new();
 
     loop {
-        let possible = current.join(".cargo").join("config");
-        if fs::metadata(&possible).is_ok() {
-            let file = try!(File::open(&possible));
+        let possible = ConfigFile::new(current.join(".cargo").join("config"));
+        if possible.exist() && stash.get(&possible).is_none() {
```

Comment:
Instead of `get` + `insert` you can actually just use `insert` here and check the boolean return value to see if it's the first time it was inserted.